Base VHDS implementation

dmitri-d opened the pull request:
For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

dmitri-d commented:
@htuch, @brian-avery: This is the base (no on-demand updates) implementation of VHDS. Please note that RouteConfiguration.virtual_hosts and vhds can be used together now.

dmitri-d commented:
* rebased.

brian-avery commented:
LGTM

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -159,21 +179,117 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+  }
+}
+
+void VhdsSubscription::rebuild_route_config(
```
Thanks for this contribution, this will be foundational to VHDS.

I'm going to wait until the API PR is merged before diving into this in depth, but as a starting point, can you please make sure things like formatting and naming match Envoy convention and style guide https://github.com/envoyproxy/envoy/blob/master/STYLE.md. E.g. this should be `rebuildRouteConfig`, below should be `for (const auto` etc. Running `fix_format` will help with a bunch.

Don't want to be too nit picky, but it will make the review easier to focus on the substance if you can clean up the style issues in advance.

dmitri-d commented:
* rebased
* fixed formatting issues
* added VHDS config validation tests

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -116,10 +117,23 @@ void RdsRouteConfigSubscription::onConfigUpdate(
-    for (auto* provider : route_config_providers_) {
-      provider->onConfigUpdate();
+
+    if (!route_config_proto_.has_vhds()) {
```
Nit: prefer to have positive vs. negative logic as the conditional in `if {} else {}`

dmitri-d replied:
fixed

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -116,10 +117,23 @@ void RdsRouteConfigSubscription::onConfigUpdate(
+      }
+      vhds_subscription_.release();
+    } else {
+      auto s = new VhdsSubscription(route_config_proto_, factory_context_, stat_prefix_,
```
Can you create the `std::make_unique` here? The idea is that generally in Envoy, we prefer to have smart pointers everywhere and it makes it easier to reason about cleanup and leaks if we don't need to think about whether code between this line and when we assign to the smart pointer might throw an exception etc.

dmitri-d replied:
fixed

htuch commented on include/envoy/router/router.h:
```diff
@@ -786,6 +786,8 @@ class Config {
    */
   virtual const std::string& name() const PURE;
+
+  virtual bool usesVhds() const PURE;
```
How is this used? I don't see any call sites. Is this for later PRs?

dmitri-d replied:
Yes, it's used in on-demand VHDS. I can remove it here if you prefer.

htuch replied:
Yeah, could be good to punt to that.

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -116,10 +117,23 @@ void RdsRouteConfigSubscription::onConfigUpdate(
+                                    route_config_providers_);
+      s->registerInitTargetWithInitManager(factory_context_.initManager());
+      vhds_subscription_.reset(s);
+      last_update_details_ = std::make_unique<VhdsConfigUpdateDetails>(*vhds_subscription_);
```
I'd probably find it a little clearer as `config_update_info` (and `ConfigUpdateInfo` for types), but this is just, like, my opinion :) It did take me a few minutes to figure out what this was all about.

dmitri-d replied:
done

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -116,10 +117,23 @@ void RdsRouteConfigSubscription::onConfigUpdate(
+      ENVOY_LOG(debug, "rds: loading new configuration: config_name={} hash={}", route_config_name_,
+                new_hash);
+
+      last_update_details_ = std::make_unique<RdsConfigUpdateDetails>(route_config_proto_,
```
Would having VHDS and RDS details distinct vs. merged provide any benefit? I'm thinking specifically about the idea we once had of allowing the config dump to be used to reconstruct the original configuration. It's a bit hokey having to plumb route config info down to VHDS to provide config dump info, but it works.

dmitri-d replied:
It's a bit more involved: ```VhdsSubscription``` uses RouteConfiguration it receives from the parent ```RdRdsRouteConfigSubscription``` to build a complete RouteConfiguration (rds + vhds). ```RdsRouteConfigProviderImpl``` uses the complete RouteConfiguration then to update worker threads.

I like the idea of merging RDS and VHDS update details into one class. We can then pass an instance of it into ```RouteConfigProvider.onConfigUpdate()``` (https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.h#L155), instead of ```RdsRouteConfigProvider``` reaching into ```RdRdsRouteConfigSubscription```. In other words to use it both for updating RouteConfig in worker threads and for generation of config dumps. 

What do you think?

htuch replied:
Sounds like a nice potential cleanup. We could punt on it to a followup PR though in the interest of reducing churn or do it here as you prefer.

dmitri-d replied:
I think I'd rather do it here, tbh.

dmitri-d replied:
Done, see https://github.com/envoyproxy/envoy/pull/6438/files#diff-7ccfbe677293512b58aac30fa4e8c26a.

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -159,21 +173,121 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
       [this, new_config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = new_config; });
 }
 
+VhdsSubscription::VhdsSubscription(
```
Should this be split into its own file? I know this entire file has grown to be a monolith, so fine to just punt with a TODO, but I'd like to see it become more decomposed.

dmitri-d replied:
Looks like it's doable. I'll give it a try.

dmitri-d replied:
Done. I extended ```RdsConfigProvider``` interface with ```onConfigUpdate``` method (https://github.com/envoyproxy/envoy/pull/6438/files#diff-328729732378ad92588209ae87b94b78R47), but can add another interface just for that, or provide a default implementation right there.

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -159,21 +173,121 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+  }
+
+  initializeVhosts(route_config_proto_);
+  subscription_ = factory_function(
```
@fredlas could you provide a review of the delta xDS parts? Thanks!

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -159,21 +173,121 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+void VhdsSubscription::updateVhosts(
+    std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources) {
+  // TODO (dmitri-d) validate added_resources?
```
Yes, we should do this!

dmitri-d replied:
Added virtual host validation.

htuch commented on source/common/router/rds_impl.cc:
```diff
@@ -159,21 +173,121 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+
+void VhdsSubscription::initializeVhosts(
+    const envoy::api::v2::RouteConfiguration& route_configuration) {
+  if (route_configuration.virtual_hosts_size() <= 0) {
```
Doesn't the `for` loop below basically do this?

dmitri-d replied:
Hahaha. Yes, it does indeed. Fixed.

htuch commented on source/common/router/rds_impl.h:
```diff
@@ -135,12 +196,74 @@ class RdsRouteConfigSubscription : Envoy::Config::SubscriptionCallbacks,
+  absl::optional<LastConfigInfo> config_info_;
+};
+
+struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
```
Why is this now public?

dmitri-d replied:
Made is private again. An artefact from some previous iteration of VHDS implementation.

fredlas commented on source/common/router/vhds.h:
```diff
@@ -0,0 +1,100 @@
+  Stats::ScopePtr scope_;
+  VhdsStats stats_;
+  TimeSource& time_source_;
+  SystemTime last_updated_;
```
It looks like last_updated_ and lastUpdated() aren't used for anything. Did I miss a use, or could it be removed?

dmitri-d replied:
These are used here: https://github.com/envoyproxy/envoy/pull/6438/files#diff-a112a0982616fac9024fc1c8e42f0661R121. Please also see the thread at https://github.com/envoyproxy/envoy/pull/6438#discussion_r274124191.

dmitri-d commented on source/common/router/vhds.h:
```diff
@@ -0,0 +1,100 @@
+        Upstream::ClusterManager&, Envoy::Runtime::RandomGenerator&, Stats::Scope&, const std::string&,
+        const std::string&, absl::string_view, Api::Api&);
+
+class VhdsSubscription : Envoy::Config::SubscriptionCallbacks,
```
Added (please let me know if you had something more substantial in mind).

htuch commented on test/integration/vhds_integration_test.cc:
```diff
@@ -0,0 +1,298 @@
+namespace Envoy {
+namespace {
+
+const char Config[] = R"EOF(
```
I still hold out hope that we can integrate cleanly tests like this with the https://github.com/envoyproxy/envoy/blob/master/test/config/utility.h approach, but not sure how realistic. CC @fredlas @alyssawilk

htuch replied:
Ping on this one.

htuch commented on test/integration/vhds_integration_test.cc:
```diff
@@ -0,0 +1,298 @@
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(VhdsIntegrationTest, RdsWithVirtualHostsVhdsVirtualHostAddUpdateRemove) {
```
Can you add single line `//` human readable description above each `TEST_P`?

dmitri-d replied:
Added

dmitri-d commented on source/common/router/vhds.cc:
```diff
@@ -0,0 +1,120 @@
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/config/rds_json.h"
```
removed.

htuch replied:
@dmitri-d could you click "Resolve conversation" when a thread is handled? That makes review a bit easier. Thanks :)

dmitri-d commented on source/common/router/rds_impl.cc:
```diff
@@ -116,10 +117,22 @@ void RdsRouteConfigSubscription::onConfigUpdate(
+      vhds_subscription_->registerInitTargetWithInitManager(factory_context_.initManager());
+      config_update_info_ = std::make_unique<VhdsConfigUpdateInfo>(*vhds_subscription_);
+    } else {
+      ENVOY_LOG(debug, "rds: loading new configuration: config_name={} hash={}", route_config_name_,
```
Added.

dmitri-d commented:
- Merged details of RDS and VHDS updates and extracted resulting functionality into a dedicated class. 
- RouteConfigProviderManagerImpl, RdsRouteConfigProviderImpl, and RdsRouteConfigSubscription can be decoupled now (probably a job for a different PR though)
- merged upstream changes

Docs/comments are still outstanding, but if these changes are ok, I'll get on that right away.

dmitri-d commented:
Oh, and I wonder if ```RouteConfigProvider::ConfigInfo``` should be moved out out of ```RouteConfigProvider```. I can argue both ways...

dmitri-d commented:
@htuch: resolved a bunch of threads.

htuch commented on source/common/router/rds_impl.h:
```diff
@@ -98,6 +103,11 @@ class RdsRouteConfigSubscription : Envoy::Config::SubscriptionCallbacks,
+  std::unordered_set<RouteConfigProvider*>& routeConfigProviders() {
+    return route_config_providers_;
+  }
+  RouteConfigUpdatePtr& routeConfigUpdateInfo() { return config_update_info_; }
```
Inconsistency here between the type "RouteConfigUpdate" and "routeConfigUpdateInfo".

htuch commented on source/common/router/route_config_update_impl.cc:
```diff
@@ -0,0 +1,98 @@
+  route_config_proto_ = rc;
+  last_config_hash_ = new_hash;
+  last_config_version_ = version_info;
+  last_updated_ = start_of_update;
```
Could just inline the `start_of_update` definition here, this doesn't need to be precise.

htuch commented on source/common/router/route_config_update_impl.cc:
```diff
@@ -0,0 +1,98 @@
+  updateVhosts(virtual_hosts_, added_resources);
+  rebuildRouteConfig(virtual_hosts_, route_config_proto_);
+
+  const uint64_t new_hash = MessageUtil::hash(route_config_proto_);
```
Can you just call into `onRdsUpdate` here?

htuch commented on source/common/router/route_config_update_impl.h:
```diff
@@ -0,0 +1,53 @@
+public:
+  RouteConfigUpdateImpl(TimeSource& time_source) : time_source_(time_source) {}
+
+  bool onRdsUpdate(const envoy::api::v2::RouteConfiguration& rc,
```
For the overrides, we typically group them separated by a blank line and then put `// Router::RouteConfigupdate` or something like that above. E.g. see https://github.com/envoyproxy/envoy/blob/3abb7d051c55f190d19b531a2fd1360842b334a5/source/common/http/conn_manager_impl.h#L66 for an example.

dmitri-d commented:
@htuch: added missing docs, merged latest changes.

htuch commented on include/envoy/router/route_config_update_receiver.h:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Called on updates via VHDS.
+   * @param added_resources supplies Resources (each containing a VirtualHost) that have been
+   * added..
```
Nit: `..`

htuch commented on include/envoy/router/route_config_update_receiver.h:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * @return std::string& the version of RouteConfiguration.
+   */
+  virtual std::string& configVersion() PURE;
```
Why isn't this `const`?

htuch commented on include/envoy/router/route_config_update_receiver.h:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * @return envoy::api::v2::RouteConfiguration& current RouteConfiguration.
+   */
+  virtual envoy::api::v2::RouteConfiguration& routeConfiguration() PURE;
```
Why isn't this `const` as well?

dmitri-d commented:
/retest

dmitri-d commented:
@htuch: looks like coverage test passes (it core dumped on a previous run). release and ipv6 test are failing, and I think this isn't a flake, ads_integration_test seems to be passing everywhere else. For some reason rate limit test time outs when running on ipv6. Is there any way to run just this test suite on the CI with trace-level logging (tests pass when I run them on a local vm)? 

```
[ RUN      ] IpVersionsClientType/AdsIntegrationTest.BasicWithRateLimiting/0
[2019-05-09 22:56:53.919][21433][critical][assert] [test/integration/http_integration.cc:313] assert failure: result. Details: Timed out waiting for new connection.
[2019-05-09 22:56:53.919][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Aborted, suspect faulting address 0x53b9
[2019-05-09 22:56:53.919][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-05-09 22:56:53.926][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #0: Envoy::SignalAction::sigHandler() [0x242a8ec]
[2019-05-09 22:56:53.926][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #1: __restore_rt [0x7efc80e5b390]
[2019-05-09 22:56:53.933][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #2: Envoy::HttpIntegrationTest::waitForNextUpstreamRequest() [0xef49a7]
[2019-05-09 22:56:53.939][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #3: Envoy::HttpIntegrationTest::sendRequestAndWaitForResponse() [0xef485b]
[2019-05-09 22:56:53.946][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #4: Envoy::HttpIntegrationTest::makeHeaderOnlyRequest() [0xef677c]
[2019-05-09 22:56:53.952][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #5: Envoy::HttpIntegrationTest::testRouterHeaderOnlyRequestAndResponse() [0xef6987]
[2019-05-09 22:56:53.958][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #6: Envoy::(anonymous namespace)::AdsIntegrationTest::makeSingleRequest() [0xe932dd]
[2019-05-09 22:56:53.965][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #7: Envoy::(anonymous namespace)::AdsIntegrationTest::testBasicFlow() [0xe92150]
[2019-05-09 22:56:53.971][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #8: Envoy::(anonymous namespace)::AdsIntegrationTest_BasicWithRateLimiting_Test::TestBody() [0xe93d20]
[2019-05-09 22:56:53.978][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #9: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x280f08e]
[2019-05-09 22:56:53.984][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #10: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x280038e]
[2019-05-09 22:56:53.990][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #11: testing::Test::Run() [0x27eaf33]
[2019-05-09 22:56:53.997][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #12: testing::TestInfo::Run() [0x27ebb38]
[2019-05-09 22:56:54.003][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #13: testing::TestSuite::Run() [0x27ec213]
[2019-05-09 22:56:54.009][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #14: testing::internal::UnitTestImpl::RunAllTests() [0x27f965a]
[2019-05-09 22:56:54.016][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #15: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x2811f1e]
[2019-05-09 22:56:54.022][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #16: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x28027a4]
[2019-05-09 22:56:54.028][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #17: testing::UnitTest::Run() [0x27f9333]
[2019-05-09 22:56:54.034][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #18: RUN_ALL_TESTS() [0x1c8dab1]
[2019-05-09 22:56:54.041][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #19: Envoy::TestRunner::RunTests() [0x1c8d842]
[2019-05-09 22:56:54.047][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #20: main [0x1c8c694]
[2019-05-09 22:56:54.047][21433][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #21: [0x7efc801f3830]
external/bazel_tools/tools/test/test-setup.sh: line 310: 21433 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1

```

htuch commented:
@dmitri-d coverage looks good now. One thing you could do is merge master, there have been a bunch of stability fixes to CI since you last did this.

In terms of traces and running tests, you can hack https://github.com/envoyproxy/envoy/blob/master/ci/do_ci.sh to run whatever Bazel commands you want, including limiting the test filter and increasing trace level, for the CI targets you care about, on your branch and CI will run it.

dmitri-d commented:
/retest

dmitri-d commented:
* rebased
* squashed
* signed-off
* upstream/downstream connections get cleaned up in ads_integration_test teardown: https://github.com/envoyproxy/envoy/pull/6438/files#diff-158efafdef83dc24db33be613e278d33R69. Without it I was still seeing timeouts (weirdly, only when all tests were being run. When running just integration tests, ads_integration_test was passing without any changes).

dmitri-d commented:
@htuch: ci appears to be happy now.

dmitri-d commented:
Thanks for reviews, @htuch!

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -111,15 +109,23 @@ void RdsRouteConfigSubscription::onConfigUpdate(
+                route_config_name_, config_update_info_->configHash());
+      vhds_subscription_ = std::make_unique<VhdsSubscription>(
+          config_update_info_, factory_context_, stat_prefix_, route_config_providers_);
+      vhds_subscription_->registerInitTargetWithInitManager(factory_context_.initManager());
```
is there any thought around management server sends RouteConfiguration with vhds after the initManager has been initialized?
